src/robot/running/keywordrunner.py

```diff
@@ -179,19 +197,54 @@ def _get_values_for_one_round(self, data):
 
     def _replace_variables(self, data):
         values = self._context.variables.replace_list(data.values)
-        if data.range:
-            values = self._get_range_items(values)
-        if len(values) % len(data.variables) == 0:
+        values = self._transform_items(values)
+        values_per_iteration = self._original_values_per_iteration(data.variables)
+        if len(values) % values_per_iteration == 0:
             return values
-        raise DataError('Number of FOR loop values should be multiple of '
-                        'variables. Got %d variables but %d value%s.'
-                        % (len(data.variables), len(values), s(values)))
+        raise DataError(self._replace_variables_error_message(data, values))
+
+    def _replace_variables_error_message(self, data, values):
+        return ('Number of FOR loop values should be multiple of '
+                + 'its variables. Got %d variables but %d value%s.'
```

Comment:
`+` isn't actually needed here because Python implicitly concatenates strings like `"foo" "bar"`. That is especially handy when doing string formatting because

```python
("A: %s\n"
 "B: %s" % (1, 2))
```
works but
```python
("A: %s\n"
 + "B: %s" % (1, 2))
```
does not.